### ISSUE TITLE
Syncing live_update with multiple / derivative images?

### ISSUE THREAD

**Issue by Kezzsim** [tilt-dev/tilt#3502]
I have a Loopback/Express.js project that builds one container as a base with all methods specified:
(_./app/Dockerfile_)
```
FROM node:10-slim
COPY --chown=node . .
## other build instructions ##
```
And then builds derivative images by calling that as the baseimage, with the contents of the `server` folder overrided with different configurations from that context:
(_./subservice1/Dockerfile_)
```
FROM thisproject_core:latest
COPY --chown=node . ./server
## other build instructions ##
```
Combined into a docker-compose.yml, the result is something like this, where _core_ is the first Dockerfile and subservice1 & 2 are products of the later:
```
  core:
    image: thisproject_core
    build: './app'
    container_name: core
  subservice1:
    image: thisproject_subservice1
    build: './subservice1'
    container_name: subservice1
  subservice2:
    image: thisproject_subservice2
    build: './subservice1'
    container_name: subservice2
```
This is all fine and dandy and I can add a `live_update` sync path to watch the subservice folders with no problems. However, what if I need multiple sync paths? Because each time I update the main _core_ image, the subservices should pull in those changes and restart.

I've tried doing this:
```
docker_compose("./docker-compose.yml")
docker_build('thisproject_core', './app', live_update = [
    sync('./app', '/home/node/app'),
    run('npm i', trigger='package.json'),
    restart_container()
])
docker_build('thisproject_subservice1', './subservice1', live_update = [
    sync('./app', '/home/node/app'),
    sync('./subservice1', '/home/node/app/server'),
    restart_container()
])
```
Such that both `docker_builds` are watching ./app for changes... but that results in this error:
```
sync step source '/Users/kezz/Desktop/thisproject/app' is not a child of any watched filepaths ([/Users/kezz/Desktop/thisproject/subservice1])
```
Is the implication that the sync path is limited by docker's build context? I mean, that's alright but then I have to find a way to rebuild the core image whenever I make a change so the derivative containers can pick it up... 
Let me know if there's any known workarounds for this, it's really cool that Docker can re-use images to cut down on resources when building similar replicas, it would be great to tap into that with tilt!

**Comment by landism** [comment]
> Is the implication that the sync path is limited by docker's build context?

Kind of, but really more like "limited by the build contexts of the image and all of its ancestor images. What you want to do should be possible.

If I understand your question correctly, this is a simple example of what you want that works for me.
a/Dockerfile:
```
FROM alpine:latest
ADD test.txt /
```
b/Dockerfile:
```
FROM a
ADD go.sh /
ENTRYPOINT ./go.sh
```

Tiltfile:
```
docker_build('a', 'a')
docker_build('b', 'b', live_update=[
  sync('a', '/'),
  sync('b', '/'),
])
k8s_yaml('deploy.yaml')
```

I don't see an obvious problem with the setup you described. A few avenues for exploration:
* Maybe Tilt isn't making the connection between the core image and the subservice image? When you make a change to the app directory, does it rebuild both core and subservice?
* If you change your docker-compose.yaml so that it only has subservice1 and no core or subservice2, does Tilt generate a warning that the core image is unused?
* The report says your subservice1 Dockerfile is at ./pods/subservice1/Dockerfile and your core Dockerfile is at ./app/Dockerfile, but subservice1's syncs are for './app' and './subservice1', which seems inconsistent. Is this a transcription error or are things more complicated?

**Comment by nicks** [comment]
For what it's worth, here's an example of this pattern in our integration tests:
https://github.com/tilt-dev/tilt/blob/master/integration/live_update_base_image/Tiltfile

+1 on the rest of @landism 's comment, I don't see an obvious reason why this wouldn't work.

**Comment by Kezzsim** [comment]
@landism you're right about that last bit being a transcription error, I've fixed it in the original comment.
The good news is I figured out what was causing tilt not to link the images - I had left the old image name that docker-compose originally assigned to the core image in the `FROM` tag of the two subservices... Once I fixed that I was able to run my tiltfile with the live updates as described above.

Unfortunately that was not the end of my issues with this pattern, Tilt has been doing what I can only describe as 'whack-a-mole' ever since, where _core_ will spin up, and then _subservice1_ but when _subservice2_ spins up, without any interaction from the user, it forces _core_ and _subservice1_ to spontaneously rebuild and it gets caught in a loop of doing that over and over again...

This behavior persists even after:
- Removing the `live_update' portion of both subservices
- calling `docker system prune` to remove all images and then regenerate them

It stops when I remove the `docker_build`s from the tiltfile for the subservices... but then I still have the problem of not being able to use the live update feature. I'm not using volumes or any other thing that could link container contents to a location on the machine so I'm confused at how the rebuild is getting triggered automatically without me doing anything.

The only way to get everything stable right now is to remove almost everything from my tiltfile, except the `docker-compose` file and the `docker_build` for core.

Have you seen this kinda thing happen before?

**Comment by landism** [comment]
If you're on tilt v0.15.0, I'd guess that's #3496, and we just released v0.15.1 to fix it. If it's not, lemme know and we can look at other possible causes.

**Comment by Kezzsim** [comment]
@landism you're absolutely correct, it was that exact issue and upgrading to v0.15.1 resolved it immediately.
It's able to watch the parent and child image containers perfectly and sync the changes to all of them, without having the rebuild the container.

Fantastic work, thank you all very much!
